Allow Iceberg materialized views to federate

findepi opened the pull request:
Probably fixes https://github.com/trinodb/trino/issues/13131

ebyhr reviewed (commented):
Looks good to me. I leave the approval to other reviewers who're familiar with materialized views.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -2453,9 +2458,19 @@ public MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession s
+        Iterable<String> tableToSnapshotIds = Splitter.on(',').split(dependsOnTables);
+        for (String entry : tableToSnapshotIds) {
+            if (entry.equals(UNKNOWN_SNAPSHOT_TOKEN)) {
+                // This is a federated materialized view. Trust user's choice and assume "fresh or fresh enough".
```
changed to 
```
// This is a "federated" materialized view (spanning across connectors). Trust user's choice and assume "fresh or fresh enough".
```

findepi commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:
```diff
@@ -1119,6 +1124,35 @@ public void testMaterializedView()
+        String query = "SELECT node_version, count(*) AS count FROM system.runtime.nodes GROUP BY 1";
+        String create = "CREATE MATERIALIZED VIEW " + viewName + " AS " + query;
+        if (!hasBehavior(SUPPORTS_CREATE_FEDERATED_MATERIALIZED_VIEW)) {
+            assertQueryFails(create, "This connector does not support creating materialized views");
```
Yes, it's intentional.
Note that previous behavior -- allow CREATE but don't allow REFRESH -- was useless, i.e. a bug. Such a mview didn't serve any purpose other than a regular view.

> Also, the failure message is misleading when the connector supports creating non-federated mv.

When we have such a case, we will update this. Will add a Note comment.

findepi commented:
thank you @ebyhr for your review. AC. PTAL

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -2453,9 +2458,19 @@ public MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession s
+        Iterable<String> tableToSnapshotIds = Splitter.on(',').split(dependsOnTables);
+        for (String entry : tableToSnapshotIds) {
+            if (entry.equals(UNKNOWN_SNAPSHOT_TOKEN)) {
+                // This is a "federated" materialized view (spanning across connectors). Trust user's choice and assume "fresh or fresh enough".
```
Once a storage table is created from non-iceberg sources, does this mean that subsequent REFRESHes will always be a no-op since the MV is always "fresh" ? If so, is there some way to force refresh for such an MV ?

findepi replied:
Good catch. I've added test coverage for that. Not solved yet.

findepi commented:
@raunaqmorarka added tests reproducing the problem and addressed

We have effectively three different states an MV can be in

- "fresh" -- can be queried, and REFRESH is a no-op
- "stale" -- cannot be queried (needs to be inlined), and REFRESH should refresh it
- "in between" -- e.g. federated mviews -- can be queried, but REFRESH should refresh it

To support this, i changed `MaterializedViewFreshness`. It's `boolean` field could not convey those different meanings, so replaced it with an enum.

cc @losipiuk @sopel39  @martint  @alexjo2144 @romanvainb

findinpath commented on core/trino-main/src/test/java/io/trino/metadata/TestInformationSchemaMetadata.java:
```diff
@@ -38,6 +38,8 @@
 import io.trino.testing.LocalQueryRunner;
 import io.trino.transaction.TransactionId;
 import io.trino.transaction.TransactionManager;
```
Same as you did already on  d57d6602 it would be clearer to some why doing the resource allocation within the test constructor is a bad idea.

> Maven surefire plugin does not report full exception when it happens during test class constructor call.

findepi replied:
I am automating checks now: https://github.com/trinodb/trino/pull/15165
So we can improve comments / error reporting in ReportResourceHungryTests in that PR.

findinpath commented on testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:
```diff
@@ -122,9 +122,15 @@ protected QueryRunner createQueryRunner()
-                                new SchemaTableName("default", "tests_table"),
-                                new SchemaTableName("tiny", "nation")))
+                        .withListTables((session, schemaName) -> {
+                            if (schemaName.equals("default")) {
```
```
                            return switch (schemaName) {
                                case "default" -> List.of("tests_table");
                                case "tiny" -> List.of("nation");
                                default -> List.of();
                            };
```

findinpath commented on testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:
```diff
@@ -122,9 +122,15 @@ protected QueryRunner createQueryRunner()
-                                new SchemaTableName("default", "tests_table"),
-                                new SchemaTableName("tiny", "nation")))
+                        .withListTables((session, schemaName) -> {
+                            if (schemaName.equals("default")) {
```
nit: In some methods is used `equalsIgnoreCase` and some methods use `equals`.
It would be good to have this kind of checks consistent across the code base.

findepi replied:
I agree. I think i didn't introduce any new equalsIgnoreCase and didn't remove any existing equalsIgnoreCase occurrences though.
This would be suboptimal to mix two different refactors together.

findinpath commented on core/trino-main/src/main/java/io/trino/connector/system/MaterializedViewSystemTable.java:
```diff
@@ -57,7 +56,7 @@
             .column("storage_schema", createUnboundedVarcharType())
             .column("storage_table", createUnboundedVarcharType())
-            .column("is_fresh", BOOLEAN)
+            .column("freshness", createUnboundedVarcharType())
```
Naive question: Can we change freely without worrying of breaking changes the system tables?

findepi replied:
Every change like this can break something. However, I don't see any better way to approach that.
Keeping `is_fresh` column as deprecated would force me to determine whether UNKNOWN freshness is "true" or "false", but it's neither. This could also break tools.
Explicit breakage is often better than a hidden one.

findinpath commented on core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:
```diff
@@ -937,13 +938,6 @@ public InsertTableHandle beginRefreshMaterializedView(Session session, TableHand
         sourceConnectorHandles.add(tableHandle.getConnectorHandle());
 
-        if (sourceConnectorHandles.stream()
-                .map(Object::getClass)
```
👍

findepi replied:
thank you

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -2356,8 +2360,12 @@ public Optional<ConnectorOutputMetadata> finishRefreshMaterializedView(
         String dependencies = sourceTableHandles.stream()
-                .map(handle -> (IcebergTableHandle) handle)
-                .map(handle -> handle.getSchemaTableName() + "=" + handle.getSnapshotId().map(Object.class::cast).orElse(""))
+                .map(handle -> {
```
Currently the freshness is considered based on Iceberg tables.
However historization of the changes is done for Delta tables as well.
Just something to consider doing at the engine level.

findepi replied:
Very good point. This might get addressed in https://github.com/trinodb/trino/pull/12499

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -2471,17 +2491,17 @@ else if (strings.size() != 2) {
             }
         }
-        return new MaterializedViewFreshness(true);
+        return new MaterializedViewFreshness(hasUnknownTables ? UNKNOWN : FRESH);
```
Can you pls add a test assertion showcasing `UNKNOWN` freshness state?

findinpath commented:
```
testing/bin/ptl env up --environment 'singlenode'  --config 'config-default' --without-trino
```


```
create table iceberg.default.itable1 (x integer);

insert into iceberg.default.itable1 values (1),(2),(3);

create materialized view iceberg.default.mvtest1 as select nationkey, name from iceberg.default.itable1 itable1 inner join tpch.sf1.nation tnation on itable1.x = tnation.nationkey;


refresh maerialized view  iceberg.default.mvtest1;

select * from  iceberg.default.mvtest1 ;

insert into iceberg.default.itable1 values (4);

select * from  iceberg.default.mvtest1 ;
```

I see the country corresponding to the nationkey `4` when selecting from `iceberg.default.mvtest1` even though the materialized view was not refreshed. Is this supposed to be happening?


```
SELECT * FROM system.metadata.materialized_views;
```

The freshness for `mvtest1` is always `STALE` even right after refreshing the materialized view. Wasn't is supposed to be unknown due to the fact that it is "federated" with a `tpch` table?

findepi commented on testing/trino-tests/src/test/java/io/trino/connector/informationschema/BenchmarkInformationSchema.java:
```diff
@@ -142,6 +143,20 @@ public MaterializedResult queryInformationSchema(BenchmarkData benchmarkData)
         return benchmarkData.queryRunner.execute(benchmarkData.query);
     }
 
```
Good catch! Won't take care of these, but feel free to do this.

findepi commented:
(just rebased)

findepi commented:
> I see the country corresponding to the nationkey `4` when selecting from `iceberg.default.mvtest1` even though the materialized view was not refreshed. Is this supposed to be happening?

Yes.

In this case, there are two base tables

1. iceberg table 
2. non-iceberg table (tpch)

When you have a modification in the iceberg table, the MV is _known to be stale_, so the freshness is STALE.
A STALE MV gets inlined by the engine. This is the same behavior that existed before my changes.

findepi commented:
> The freshness for `mvtest1` is always `STALE` even right after refreshing the materialized view. Wasn't is supposed to be unknown due to the fact that it is "federated" with a `tpch` table?

can this be related to a typo in the command:

```
refresh maerialized view  iceberg.default.mvtest1;
```
?

otherwise I couldn't reproduce this.

findepi commented:
AC, thank you @findinpath for thorough review.

raunaqmorarka commented:
We should probably update iceberg or MV docs to explain what REFRESH MV does when all the source tables are iceberg tables vs when there is a non-iceberg table involved.

findepi commented:
@raunaqmorarka good point, updated the docs now.

@mosabua i am going to merge this when build passes. if you have comments, i will happily apply them, but it's also ok to improve the wording as a follow-up.

raunaqmorarka commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -1236,12 +1236,19 @@ used to specify the schema where the storage table will be created.
 table is up to date. If the data is outdated, the materialized view behaves
 like a normal view, and the data is queried directly from the base tables.
+Detecting outdated data is possible only when the materialized view uses
+Iceberg tables only, or when it uses mix of Iceberg and non-Iceceberg tables
```
`Iceceberg` typo

findepi replied:
thank you

mosabua commented:
So AWESOME @findepi

mosabua commented:
Filed https://github.com/trinodb/trino/issues/15199 to follow up on the docs with some more clarifications.

findinpath commented:
> can this be related to a typo in the command:

> `refresh maerialized view  iceberg.default.mvtest1;`

Indeed. It was apparently a typo. 
I've retested the scenario now and see that the freshness is set to "UNKNOWN" (due to the dependency on the non versioned `tpch` table).
